data/impl/StoreValidator.js

```diff
@@ -0,0 +1,187 @@
+
+import {ValidationState} from '../validation/ValidationState';
+
+/**
+ * Computes and stores validation state for a Store and its modified Records.
+ *
+ * @private
+ */
+export class StoreValidator extends HoistBase {
+
+    /** @member {Store} */
+    store;
+
+    @managed _validationTask = new PendingTaskModel({mode: 'all'});
+
+    // Stores the result of evaluating each record. Errors are store in a deep map structure
+    // to facilitate access at the record, field and rule levels:
+    // e.g. _errors[record.id][field.name][rule] = errors[]
+    @observable.ref _errors = {};
+
+    /** @return {boolean} */
+    @computed
+    get isValid() {
+        return this.validationState === ValidationState.Valid;
+    }
+
+    /** @return {ValidationState} - the current validation state of the store. */
+    @computed
+    get validationState() {
+        return this.getValidationState();
+    }
+
+    /** @return {boolean} - true if currently recomputing validation state. */
+    get isPending() {
+        return this._validationTask.isPending;
+    }
+
+    /** @return {string[]} */
+    @computed
+    get errors() {
+        return this.getErrors();
+    }
+
+    /** @return {number} - count of all validation errors for the store. */
+    @computed
+    get errorCount() {
+        return this.errors.length;
+    }
+
+    /**
+     * @param {Store} store - Parent store to which this validator belongs
+     */
+    constructor(store) {
+        super();
+        makeObservable(this);
+        this.store = store;
+    }
+
+    //---------------------------------------
+    // Validating Store
+    //---------------------------------------
+    /**
+     * Recompute validations for all records and return true if the store is valid.
+     * @returns {Promise<boolean>}
+     */
+    async validateAsync() {
+        if (!this.store.isModified) return true;
+        const promises = map(this.store.modifiedRecords, record => this.validateRecordAsync(record));
+        await Promise.all(promises);
+        return this.isValid;
+    }
+
+    /** @return {ValidationState} - the current validation state of the store. */
+    getValidationState() {
+        const VS = ValidationState;
+        if (!this.store.isModified) return VS.Valid;
+
+        const states = map(this.store.modifiedRecords, record => record.validationState);
+        return states.includes(VS.NotValid) ? VS.NotValid : VS.Valid;
+    }
+
+    /** @return {string[]} - list of all errors in the store */
+    getErrors() {
+        return flatten(map(this.store.modifiedRecords, record => this.getErrorsForRecord(record)));
+    }
+
+    //---------------------------------------
+    // Validating Records
+    //---------------------------------------
+    /**
+     * Recompute validations for a given record and return true if the record is valid.
+     * @returns {Promise<boolean>}
+     */
+    async validateRecordAsync(record) {
```

Comment:
Thinking a RecordValidator object, that gets created on the fly with a pointer back to the Record will clean this up a lot!  Breaks up this file (which is looping on three different levels) and also means you don't have to pass the record to all of the methods below.  Should not be too much object creation, now that we are only validating modified records.